p4_symbolic/ir/ir.cc

```diff
@@ -0,0 +1,295 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(babman): Make error messages include more context (e.g. the value
+//               of the unsupported expression).
+
+#include "p4_symbolic/ir/ir.h"
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "google/protobuf/struct.pb.h"
+
+namespace p4_symbolic {
+namespace ir {
+
+// Parsing and validating Headers.
+absl::Status ValidateHeaderTypeFields(const google::protobuf::ListValue &list) {
+  // Size must be 3.
+  int size = list.values_size();
+  if (size != 3) {
+    return absl::Status(absl::StatusCode::kInvalidArgument,
+                        "Headertype fields are badly formatted!");
+  }
+
+  // Array must contain [string, int, bool] in that order.
+  if (list.values(0).kind_case() != google::protobuf::Value::kStringValue ||
+      list.values(1).kind_case() != google::protobuf::Value::kNumberValue ||
+      list.values(2).kind_case() != google::protobuf::Value::kBoolValue) {
+    return absl::Status(absl::StatusCode::kInvalidArgument,
+                        "Headertype fields are badly formatted!");
+  }
+
+  return absl::OkStatus();
+}
+
+absl::Status TransformHeader(const bmv2::HeaderType &header,
+                             HeaderType *output) {
+  output->set_name(header.name());
+  output->set_id(header.id());
+  for (int i = 0; i < header.fields_size(); i++) {
+    const google::protobuf::ListValue &unparsed_field = header.fields(i);
+    RETURN_IF_ERROR(ValidateHeaderTypeFields(unparsed_field));
```

Comment:
Maybe rename this to ExtractHeaderFields and make it return a StatusOr<T> where T is a struct with all the the fields you need?